File: Frontrunner/etl/import_intersections.py
#!/usr/bin/env python3
"""
Import intersections from MySQL to PostgreSQL consolidated_intersections table
Groups by _location field (like "I9") and creates polygons
"""

import psycopg2
import mysql.connector
from mysql.connector import Error
import os
import re
import logging
from typing import List, Dict

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# MySQL connection config
MYSQL_CONFIG = {
    'host': os.getenv('MYSQL_HOST', 'localhost'),
    'port': int(os.getenv('MYSQL_PORT', 3306)),
    'database': os.getenv('MYSQL_DB', 'frontrunnerv3'),
    'user': os.getenv('MYSQL_USER', 'root'),
    'password': os.getenv('MYSQL_PASSWORD', '')
}

# PostgreSQL connection config
POSTGRES_CONFIG = {
    'host': os.getenv('DB_HOST', 'localhost'),
    'port': int(os.getenv('DB_PORT', 5432)),
    'database': os.getenv('DB_NAME', 'infrastructure_db'),
    'user': os.getenv('DB_USER', 'infra_user'),
    'password': os.getenv('DB_PASSWORD', 'infra_password')
}

def get_mysql_connection():
    """Get MySQL database connection"""
    try:
        conn = mysql.connector.connect(**MYSQL_CONFIG)
        return conn
    except Error as e:
        logger.error(f"Error connecting to MySQL: {e}")
        raise

def get_postgres_connection():
    """Get PostgreSQL database connection"""
    try:
        conn = psycopg2.connect(**POSTGRES_CONFIG)
        return conn
    except Exception as e:
        logger.error(f"Error connecting to PostgreSQL: {e}")
        raise

def extract_intersection_location(name: str) -> str:
    """Extract _location identifier from intersection name (e.g., 'I9' from 'I9' or 'I9_1')"""
    if not name:
        return None
    
    # Match pattern like "I9", "I10", etc.
    match = re.match(r'^([A-Z]\d+)', name)
    if match:
        return match.group(1)
    
    # If name is already just the identifier
    if re.match(r'^[A-Z]\d+$', name):
        return name
    
    return name.split('_')[0] if '_' in name else name

def extract_intersections_from_mysql() -> List[Dict]:
    """Extract intersection data from MySQL map_intersection table"""
    mysql_conn = get_mysql_connection()
    cursor = mysql_conn.cursor(dictionary=True)
    
    try:
        # First, try to get intersections from map_intersection table
        # Check if they have geometry_wkt that we can extract points from
        query1 = """
        SELECT 
            mi._OID_ as intersection_id,
            mi.name as intersection_name,
            mi._CID_ as category_type,
            mi.geometry_wkt,
            mi.is_open
        FROM map_intersection mi
        WHERE mi.name IS NOT NULL
        ORDER BY mi.name
        """
        
        cursor.execute(query1)
        intersections = cursor.fetchall()
        logger.info(f"Found {len(intersections)} intersections in map_intersection table")
        
        # If we have geometry_wkt, we can extract points from it
        # But we also need to check if they're linked to coordinates via survey_location
        query2 = """
        SELECT DISTINCT
            mi._OID_ as intersection_id,
            mi.name as intersection_name,
            mi._CID_ as category_type,
            c._OID_ as coordinate_id,
            c.latitude,
            c.longitude,
            c.altitude
        FROM map_intersection mi
        LEFT JOIN map_location ml ON mi.name = ml.name
        LEFT JOIN survey_location sl ON ml._location_survey = sl._OID_
        LEFT JOIN survey_location__shapeloc__x_y_z slsxyz ON sl._OID_ = slsxyz._OID_
        LEFT JOIN coordinate c ON c._OID_ = slsxyz._coordinate
        WHERE c.latitude IS NOT NULL 
            AND c.longitude IS NOT NULL
            AND c.latitude BETWEEN -60 AND -20
            AND c.longitude BETWEEN 100 AND 160
        ORDER BY mi.name, c._OID_
        """
        
        cursor.execute(query2)
        results = cursor.fetchall()
        
        if not results:
            logger.warning("No intersection coordinates found via survey_location. Trying alternative method...")
            # Try checking if intersections are already in locations table
            return None
        
        logger.info(f"Extracted {len(results)} intersection coordinate points from MySQL")
        return results
        
    except Error as e:
        logger.error(f"Error extracting intersections from MySQL: {e}")
        raise
    finally:
        cursor.close()
        mysql_conn.close()

def check_intersections_in_postgres() -> List[Dict]:
    """Check if intersections are already in PostgreSQL locations table"""
    conn = get_postgres_connection()
    cursor = conn.cursor()
    
    try:
        query = """
        SELECT 
            location_name,
            category,
            latitude,
            longitude,
            altitude,
            geometry_point
        FROM locations
        WHERE category = 'intersection'
            OR location_name ~ '^[A-Z]\\d+'
            OR location_name ILIKE '%intersection%'
        ORDER BY location_name
        """
        
        cursor.execute(query)
        results = cursor.fetchall()
        
        # Convert to dict format
        intersections = []
        for row in cursor.fetchall():
            intersections.append({
                'location_name': row[0],
                'category': row[1],
                'latitude': row[2],
                'longitude': row[3],
                'altitude': row[4],
                'geometry_point': row[5]
            })
        
        logger.info(f"Found {len(intersections)} intersections in PostgreSQL locations table")
        return intersections
        
    except Exception as e:
        logger.error(f"Error checking intersections in PostgreSQL: {e}")
        return []
    finally:
        cursor.close()
        conn.close()

def create_consolidated_intersections_table():
    """Create consolidated_intersections table if it doesn't exist"""
    conn = get_postgres_connection()
    conn.autocommit = True
    cursor = conn.cursor()
    
    try:
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS consolidated_intersections (
                id SERIAL PRIMARY KEY,
                intersection_name TEXT NOT NULL,
                category TEXT DEFAULT 'intersection',
                total_points INTEGER,
                center_latitude DOUBLE PRECISION,
                center_longitude DOUBLE PRECISION,
                avg_altitude DOUBLE PRECISION,
                center_point GEOMETRY(POINT, 4326),
                intersection_polygon GEOMETRY(POLYGON, 4326),
                intersection_boundary GEOMETRY(LINESTRING, 4326),
                area_sqm DOUBLE PRECISION,
                first_recorded TIMESTAMP,
                last_recorded TIMESTAMP,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
        """)
        
        # Create indexes
        cursor.execute("""
            CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_name 
            ON consolidated_intersections (intersection_name)
        """)
        cursor.execute("""
            CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_category 
            ON consolidated_intersections (category)
        """)
        cursor.execute("""
            CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_polygon 
            ON consolidated_intersections USING GIST (intersection_polygon)
        """)
        cursor.execute("""
            CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_center 
            ON consolidated_intersections USING GIST (center_point)
        """)
        
        logger.info("Created consolidated_intersections table and indexes")
        
    except Exception as e:
        logger.error(f"Error creating table: {e}")
        raise
    finally:
        cursor.close()
        conn.close()

def consolidate_intersections_from_postgres():
    """Consolidate intersections from PostgreSQL locations table"""
    conn = get_postgres_connection()
    conn.autocommit = True
    cursor = conn.cursor()
    
    try:
        # Delete existing intersections to avoid duplicates
        cursor.execute("DELETE FROM consolidated_intersections")
        logger.info("Cleared existing consolidated intersections")
        
        # Insert consolidated intersections
        cursor.execute("""
            INSERT INTO consolidated_intersections (
                intersection_name,
                category,
                total_points,
                center_latitude,
                center_longitude,
                avg_altitude,
                center_point,
                intersection_polygon,
                intersection_boundary,
                area_sqm,
                first_recorded,
                last_recorded
            )
            SELECT 
                -- Extract intersection identifier (e.g., "I9" from "I9" or "I9_point1")
                CASE 
                    WHEN location_name ~ '^[A-Z]\\d+' THEN 
                        SUBSTRING(location_name FROM '^([A-Z]\\d+)')
                    ELSE 
                        location_name
                END as intersection_name,
                COALESCE(category, 'intersection') as category,
                count(*) as total_points,
                avg(latitude) as center_latitude,
                avg(longitude) as center_longitude,
                avg(altitude) as avg_altitude,
                ST_Centroid(ST_Collect(geometry_point)) as center_point,
                ST_ConvexHull(ST_Collect(geometry_point)) as intersection_polygon,
                ST_ExteriorRing(ST_ConvexHull(ST_Collect(geometry_point))) as intersection_boundary,
                ST_Area(ST_ConvexHull(ST_Collect(geometry_point))::geography) as area_sqm,
                min(created_at) as first_recorded,
                max(created_at) as last_recorded
            FROM locations 
            WHERE (category = 'intersection'
                OR location_name ~ '^[A-Z]\\d+'
                OR location_name ILIKE '%intersection%')
                AND geometry_point IS NOT NULL
                AND latitude IS NOT NULL
                AND longitude IS NOT NULL
            GROUP BY 
                CASE 
                    WHEN location_name ~ '^[A-Z]\\d+' THEN 
                        SUBSTRING(location_name FROM '^([A-Z]\\d+)')
                    ELSE 
                        location_name
                END,
                COALESCE(category, 'intersection')
        """)
        
        count = cursor.rowcount
        logger.info(f"✅ Consolidated {count} intersections into consolidated_intersections table")
        
        # Show summary
        cursor.execute("""
            SELECT 
                count(*) as total_intersections,
                sum(total_points) as total_points,
                avg(total_points) as avg_points,
                sum(area_sqm) as total_area_sqm
            FROM consolidated_intersections
        """)
        result = cursor.fetchone()
        logger.info(f"Summary: {result[0]} intersections, {result[1]} total points, {result[2]:.1f} avg points, {result[3]:.0f} sqm total area")
        
    except Exception as e:
        logger.error(f"Error consolidating intersections: {e}")
        raise
    finally:
        cursor.close()
        conn.close()

def main():
    """Main import function"""
    logger.info("🚀 Starting intersection import...")
    
    try:
        # Step 1: Create table
        logger.info("Step 1: Creating consolidated_intersections table...")
        create_consolidated_intersections_table()
        
        # Step 2: Try to extract from MySQL
        logger.info("Step 2: Checking for intersections in MySQL...")
        mysql_intersections = extract_intersections_from_mysql()
        
        # Step 3: Consolidate from PostgreSQL locations table (if they exist there)
        logger.info("Step 3: Consolidating intersections from PostgreSQL locations table...")
        consolidate_intersections_from_postgres()
        
        logger.info("✅ Intersection import completed!")
        
    except Exception as e:
        logger.error(f"❌ Import failed: {e}")
        raise

if __name__ == '__main__':
    main()

File: Frontrunner/pages/api/import-intersections.js
const { query } = require('../../lib/mysql-database');
const { Client } = require('pg');

export default async function handler(req, res) {
  if (req.method !== 'POST' && req.method !== 'GET') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  try {
    console.log('🚀 Starting intersection import from MySQL to PostgreSQL...');

    // PostgreSQL connection - use DB_HOST if set (e.g., 'postgres' in Docker), otherwise 'localhost'
    const pgHost = process.env.DB_HOST || 'localhost';
    const pgPort = parseInt(process.env.DB_PORT || (pgHost === 'postgres' ? '5432' : '5433'));
    
    const pgClient = new Client({
      host: pgHost,
      port: parseInt(pgPort),
      database: process.env.DB_NAME || 'infrastructure_db',
      user: process.env.DB_USER || 'infra_user',
      password: process.env.DB_PASSWORD || 'infra_password'
    });

    await pgClient.connect();
    console.log('✅ Connected to PostgreSQL');

    // Step 1: Create consolidated_intersections table with all fields
    await pgClient.query(`
      CREATE TABLE IF NOT EXISTS consolidated_intersections (
        id SERIAL PRIMARY KEY,
        intersection_name TEXT NOT NULL,
        category TEXT DEFAULT 'intersection',
        total_points INTEGER,
        center_latitude DOUBLE PRECISION,
        center_longitude DOUBLE PRECISION,
        avg_altitude DOUBLE PRECISION,
        center_point GEOMETRY(POINT, 4326),
        intersection_polygon GEOMETRY(POLYGON, 4326),
        intersection_boundary GEOMETRY(LINESTRING, 4326),
        area_sqm DOUBLE PRECISION,
        all_intersection_ids TEXT[],  -- Array of all intersection _OID_ values
        all_coordinate_ids TEXT[],    -- Array of all coordinate _OID_ values
        source_tables TEXT[],         -- Array of source tables (map_intersection, pit_loc)
        first_recorded TIMESTAMP,
        last_recorded TIMESTAMP,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      )
    `);

    await pgClient.query(`
      CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_name 
      ON consolidated_intersections (intersection_name)
    `);

    await pgClient.query(`
      CREATE INDEX IF NOT EXISTS idx_consolidated_intersection_polygon 
      ON consolidated_intersections USING GIST (intersection_polygon)
    `);

    console.log('✅ Created consolidated_intersections table');

    // Step 2: Extract intersections from MySQL
    // Query pit_loc table with _CID_ = 'pit_loc_intersection' (same pattern as grouped-locations.js)
    const mysqlResult = await query(`
      SELECT DISTINCT
        pl._OID_ as intersection_id,
        pl.name as intersection_name,
        pl._CID_ as category_type,
        'pit_loc' as source_table,
        c._OID_ as coordinate_id,
        c.latitude,
        c.longitude,
        c.altitude,
        c.coord_x,
        c.coord_y,
        c.coord_z,
        c.coord_heading,
        c.coord_incl,
        c.coord_status
      FROM pit_loc pl
      INNER JOIN survey_location sl ON pl._location_survey = sl._OID_
      INNER JOIN survey_location__shapeloc__x_y_z slsxyz ON sl._OID_ = slsxyz._OID_
      INNER JOIN coordinate c ON c._OID_ = slsxyz._coordinate
      WHERE pl._CID_ = 'pit_loc_intersection'
        AND c.latitude IS NOT NULL 
        AND c.longitude IS NOT NULL
        AND c.latitude BETWEEN -60 AND -20
        AND c.longitude BETWEEN 100 AND 160
        AND pl.name IS NOT NULL
      ORDER BY pl.name, c._OID_
    `);

    console.log(`✅ Extracted ${mysqlResult.rows.length} intersection points from MySQL`);
    
    // Debug: Show sample data
    if (mysqlResult.rows.length > 0) {
      console.log('📊 Sample intersection data:', JSON.stringify(mysqlResult.rows.slice(0, 3), null, 2));
    } else {
      console.warn('⚠️ WARNING: No intersections found in MySQL!');
      console.log('🔍 Checking if pit_loc table exists and has data...');
      
      // Debug query to check what's in pit_loc
      const debugResult = await query(`
        SELECT COUNT(*) as total, 
               COUNT(CASE WHEN _CID_ = 'pit_loc_intersection' THEN 1 END) as intersections
        FROM pit_loc
      `);
      console.log('📊 pit_loc table stats:', debugResult.rows[0]);
      
      await pgClient.end();
      return res.status(200).json({ 
        success: false,
        message: 'No intersections found in MySQL',
        debug: debugResult.rows[0],
        imported: 0 
      });
    }

    // Step 3: Create staging table with all coordinate data
    await pgClient.query(`
      CREATE TABLE IF NOT EXISTS intersection_points_staging (
        id SERIAL PRIMARY KEY,
        intersection_id VARCHAR(32),
        intersection_name VARCHAR(32),
        _location VARCHAR(32),
        category VARCHAR(50),
        source_table VARCHAR(50),
        coordinate_id VARCHAR(32),
        latitude DOUBLE PRECISION,
        longitude DOUBLE PRECISION,
        altitude DOUBLE PRECISION,
        coord_x DOUBLE PRECISION,
        coord_y DOUBLE PRECISION,
        coord_z DOUBLE PRECISION,
        coord_heading DOUBLE PRECISION,
        coord_incl DOUBLE PRECISION,
        coord_status DOUBLE PRECISION,
        geometry_point GEOMETRY(POINT, 4326),
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
      )
    `);

    await pgClient.query('TRUNCATE TABLE intersection_points_staging');

    function extractLocationIdentifier(name) {
      if (!name) return null;
      const nameStr = String(name).trim();
      
      if (/^[A-Z]\d+$/.test(nameStr)) {
        return nameStr;
      }
      
      if (nameStr.includes('_')) {
        const parts = nameStr.split('_');
        const firstPart = parts[0];
        const secondPart = parts[1];
        
        if (/^[A-Z]\d+$/.test(firstPart) && /^(point|pt|coord|p)\d*$/i.test(secondPart)) {
          return firstPart;
        }
        
        return nameStr;
      }
      
      return nameStr;
    }
    
    console.log(`📊 Sample intersection names:`, mysqlResult.rows.slice(0, 5).map(r => r.intersection_name));

    // Step 4: Insert points into staging with all coordinate data
    let insertedCount = 0;
    let skippedCount = 0;
    const locationGroups = {};
    
    for (const row of mysqlResult.rows) {
      const _location = extractLocationIdentifier(row.intersection_name);
      if (!_location) {
        skippedCount++;
        console.warn(`⚠️ Skipping row with no _location extracted from name: ${row.intersection_name}`);
        continue;
      }
      
      // Track location groups
      if (!locationGroups[_location]) {
        locationGroups[_location] = 0;
      }
      locationGroups[_location]++;

      await pgClient.query(`
        INSERT INTO intersection_points_staging (
          intersection_id, intersection_name, _location, category, source_table,
          coordinate_id, latitude, longitude, altitude,
          coord_x, coord_y, coord_z, coord_heading, coord_incl, coord_status,
          geometry_point
        ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, 
                 ST_SetSRID(ST_MakePoint($8, $7), 4326))
      `, [
        row.intersection_id,
        row.intersection_name,
        _location,
        row.category_type || 'intersection',
        row.source_table || 'unknown',
        row.coordinate_id,
        row.latitude,
        row.longitude,
        row.altitude,
        row.coord_x,
        row.coord_y,
        row.coord_z,
        row.coord_heading,
        row.coord_incl,
        row.coord_status
      ]);
      insertedCount++;
    }

    console.log(`✅ Inserted ${insertedCount} points into staging table (skipped ${skippedCount})`);
    console.log(`📊 Location groups found: ${Object.keys(locationGroups).length}`, Object.keys(locationGroups).slice(0, 10));
    
    // Verify staging table has data
    const stagingCheck = await pgClient.query('SELECT COUNT(*) as count FROM intersection_points_staging');
    console.log(`📊 Staging table has ${stagingCheck.rows[0].count} rows`);
    
    if (parseInt(stagingCheck.rows[0].count) === 0) {
      await pgClient.end();
      return res.status(500).json({
        success: false,
        message: 'No data inserted into staging table',
        inserted: insertedCount,
        skipped: skippedCount
      });
    }

    // Step 5: Consolidate into polygons
    await pgClient.query('TRUNCATE TABLE consolidated_intersections');
    
    console.log('🔄 Consolidating intersections into polygons...');
    
    // First, let's check what we have in staging
    const stagingPreview = await pgClient.query(`
      SELECT _location, COUNT(*) as point_count
      FROM intersection_points_staging
      GROUP BY _location
      ORDER BY point_count DESC
      LIMIT 5
    `);
    console.log('📊 Preview of staging data:', stagingPreview.rows);

    const consolidateResult = await pgClient.query(`
      INSERT INTO consolidated_intersections (
        intersection_name, category, total_points,
        center_latitude, center_longitude, avg_altitude,
        center_point, intersection_polygon, intersection_boundary,
        area_sqm, all_intersection_ids, all_coordinate_ids, source_tables,
        first_recorded, last_recorded
      )
      SELECT 
        _location as intersection_name,
        COALESCE(category, 'intersection') as category,
        count(*) as total_points,
        avg(latitude) as center_latitude,
        avg(longitude) as center_longitude,
        avg(COALESCE(altitude, 0)) as avg_altitude,
        ST_Centroid(ST_Collect(geometry_point)) as center_point,
        ST_ConvexHull(ST_Collect(geometry_point)) as intersection_polygon,
        ST_ExteriorRing(ST_ConvexHull(ST_Collect(geometry_point))) as intersection_boundary,
        ST_Area(ST_ConvexHull(ST_Collect(geometry_point))::geography) as area_sqm,
        array_agg(DISTINCT intersection_id) FILTER (WHERE intersection_id IS NOT NULL) as all_intersection_ids,
        array_agg(DISTINCT coordinate_id) FILTER (WHERE coordinate_id IS NOT NULL) as all_coordinate_ids,
        array_agg(DISTINCT source_table) FILTER (WHERE source_table IS NOT NULL) as source_tables,
        min(created_at) as first_recorded,
        max(created_at) as last_recorded
      FROM intersection_points_staging
      WHERE geometry_point IS NOT NULL
        AND latitude IS NOT NULL
        AND longitude IS NOT NULL
        AND _location IS NOT NULL
      GROUP BY _location, COALESCE(category, 'intersection')
      HAVING count(*) >= 1
    `);

    console.log(`✅ Consolidated ${consolidateResult.rowCount} intersections`);
    
    if (consolidateResult.rowCount === 0) {
      console.error('❌ ERROR: Consolidation query returned 0 rows!');
      
      // Check what's in staging
      const stagingDetails = await pgClient.query(`
        SELECT _location, COUNT(*) as point_count, 
               COUNT(DISTINCT intersection_id) as intersection_count
        FROM intersection_points_staging
        GROUP BY _location
        ORDER BY point_count DESC
        LIMIT 10
      `);
      console.log('📊 Staging table details:', stagingDetails.rows);
      
      await pgClient.end();
      return res.status(500).json({
        success: false,
        message: 'Consolidation failed - no intersections created',
        staging_details: stagingDetails.rows,
        imported: 0
      });
    }

    // Get summary
    const summary = await pgClient.query(`
      SELECT 
        count(*) as total_intersections,
        sum(total_points) as total_points,
        avg(total_points) as avg_points,
        sum(area_sqm) as total_area_sqm
      FROM consolidated_intersections
    `);

    const summaryRow = summary.rows[0];

    // Show some examples
    const examples = await pgClient.query(`
      SELECT intersection_name, total_points, area_sqm
      FROM consolidated_intersections
      ORDER BY total_points DESC
      LIMIT 10
    `);

    await pgClient.end();

    res.status(200).json({
      success: true,
      message: 'Intersections imported successfully',
      imported: consolidateResult.rowCount,
      summary: {
        total_intersections: parseInt(summaryRow.total_intersections),
        total_points: parseInt(summaryRow.total_points || 0),
        avg_points: parseFloat(summaryRow.avg_points || 0).toFixed(1),
        total_area_sqm: parseFloat(summaryRow.total_area_sqm || 0).toFixed(0)
      },
      examples: examples.rows
    });

  } catch (error) {
    console.error('❌ Error importing intersections:', error);
    console.error('❌ Stack trace:', error.stack);
    res.status(500).json({ 
      error: 'Internal server error',
      message: error.message,
      stack: process.env.NODE_ENV === 'development' ? error.stack : undefined
    });
  }
}